File: examples/blinky/icezprog.py
#!/usr/bin/env python3
"""
icezprog.py — FPGA Programmer for ICEZero (TE0876-03 REV02)
Adapted from blackmesalabs/ice_zero_prog (Kevin Hubbard, 2017)

Uses RPi.GPIO to bit-bang SPI to the FPGA configuration PROM.
Pin mapping for TE0876-03 REV02 (from Trenz pinout.xlsx):
  CFG_DONE = GPIO5  (P11-29)
  CFG_SI   = GPIO6  (P11-31)
  CFG_SS   = GPIO12 (P11-32)
  CFG_SO   = GPIO13 (P11-33)
  CFG_SCK  = GPIO16 (P11-36)
  CFG_RST  = GPIO26 (P11-37)

Usage:  python3 icezprog.py design.bin       — program FPGA SRAM
        python3 icezprog.py design.bin flash — program QSPI Flash
        python3 icezprog.py --reset          — reset & boot from Flash
        python3 icezprog.py --id             — read Flash ID
"""

import subprocess
import sys
import time

try:
    import RPi.GPIO as GPIO
except ImportError:
    print("⚠️  RPi.GPIO not available (not on Raspberry Pi?)")
    print("   Install: sudo apt-get install python3-rpi.gpio")
    sys.exit(1)


# ── GPIO16 conflict workaround ────────────────────────────────
# googlevoicehat-soundcard overlay claims GPIO16 (sdmode/amp-enable).
# We temporarily remove it before claiming GPIO16 for FPGA programming,
# then restore it afterwards.

_OVERLAY_NAME = "googlevoicehat-soundcard"
_overlay_was_removed = False


def _overlay_loaded(name):
    """Return True if the DT overlay with given name is currently loaded."""
    try:
        result = subprocess.run(
            ["dtoverlay", "-l"],
            capture_output=True, text=True, timeout=5
        )
        return name in result.stdout
    except (FileNotFoundError, subprocess.TimeoutExpired):
        return False


def _release_conflicting_overlays():
    """Remove overlays that claim GPIO16. Returns True if any were removed."""
    global _overlay_was_removed
    if _overlay_loaded(_OVERLAY_NAME):
        print(f"⚠️  Releasing '{_OVERLAY_NAME}' overlay (holds GPIO16/CFG_SCK)...")
        r = subprocess.run(
            ["sudo", "dtoverlay", "-r", _OVERLAY_NAME],
            capture_output=True, text=True, timeout=10
        )
        if r.returncode == 0:
            _overlay_was_removed = True
            time.sleep(0.3)   # let the kernel settle
            return True
        else:
            print(f"   Could not remove overlay: {r.stderr.strip()}")
            print("   Try: sudo systemctl stop pipewire pulseaudio && sudo dtoverlay -r googlevoicehat-soundcard")
            return False
    return False


def _restore_overlays():
    """Re-add overlays that were removed."""
    if _overlay_was_removed:
        subprocess.run(
            ["sudo", "dtoverlay", _OVERLAY_NAME],
            capture_output=True, timeout=10
        )
        print(f"✅  Restored '{_OVERLAY_NAME}' overlay.")

# ── Pin Definitions (BCM numbering) ──────────────────────────
CFG_DONE = 5
CFG_SI   = 6
CFG_SS   = 12
CFG_SO   = 13
CFG_SCK  = 16
CFG_RST  = 26


class IceZeroProg:
    def __init__(self):
        GPIO.setwarnings(False)
        GPIO.cleanup()   # release any previously-claimed pins
        _release_conflicting_overlays()   # free GPIO16 if held by soundcard
        GPIO.setmode(GPIO.BCM)
        GPIO.setup(CFG_SS,  GPIO.OUT, initial=GPIO.HIGH)
        GPIO.setup(CFG_SCK, GPIO.OUT, initial=GPIO.LOW)
        GPIO.setup(CFG_SI,  GPIO.OUT, initial=GPIO.LOW)
        GPIO.setup(CFG_SO,  GPIO.IN)
        GPIO.setup(CFG_RST, GPIO.OUT, initial=GPIO.LOW)
        GPIO.setup(CFG_DONE, GPIO.IN)

    def cleanup(self):
        GPIO.cleanup()
        _restore_overlays()

    def spi_begin(self):
        GPIO.output(CFG_SS, GPIO.LOW)

    def spi_end(self):
        GPIO.output(CFG_SS, GPIO.HIGH)

    def spi_xfer(self, data, nbits=8):
        """Bit-bang SPI transfer with timing delays (matching original icotools)."""
        result = 0
        for i in range(nbits - 1, -1, -1):
            GPIO.output(CFG_SI, (data >> i) & 1)
            time.sleep(0.00001)  # 10us delay like original uwait_barrier_sync
            if GPIO.input(CFG_SO):
                result |= (1 << i)
            time.sleep(0.00001)
            GPIO.output(CFG_SCK, GPIO.HIGH)
            time.sleep(0.00001)
            GPIO.output(CFG_SCK, GPIO.LOW)
            time.sleep(0.00001)
        return result

    def fpga_reset(self):
        """Reset FPGA and wait for CDONE."""
        GPIO.output(CFG_RST, GPIO.LOW)
        time.sleep(0.002)
        GPIO.output(CFG_RST, GPIO.HIGH)
        time.sleep(0.5)
        if GPIO.input(CFG_DONE) != GPIO.HIGH:
            print("Warning: CDONE is low after reset")

    def prog_sram(self, bitfile):
        """Program FPGA SRAM from .bin file."""
        print(f"Programming FPGA SRAM: {bitfile}")
        # ICE40 SPI slave config protocol: SS_B must be LOW before releasing
        # CRESET_B so the FPGA enters SPI slave (SRAM) config mode instead of
        # booting from flash.
        GPIO.output(CFG_SS, GPIO.LOW)
        self.fpga_reset()
        # CDONE is LOW here — FPGA is waiting for bitstream data (expected)

        # 8 dummy clocks with SS asserted
        for _ in range(8):
            GPIO.output(CFG_SCK, GPIO.LOW)
            GPIO.output(CFG_SCK, GPIO.HIGH)

        with open(bitfile, 'rb') as f:
            data = f.read()

        for i, byte in enumerate(data):
            for bit in range(7, -1, -1):
                GPIO.output(CFG_SI, (byte >> bit) & 1)
                GPIO.output(CFG_SCK, GPIO.LOW)
                GPIO.output(CFG_SCK, GPIO.HIGH)

        # 49 dummy clocks — CDONE should go HIGH during these if config OK
        for _ in range(49):
            GPIO.output(CFG_SCK, GPIO.LOW)
            GPIO.output(CFG_SCK, GPIO.HIGH)

        GPIO.output(CFG_SS, GPIO.HIGH)
        time.sleep(0.002)
        if GPIO.input(CFG_DONE) == GPIO.HIGH:
            print(f"✅ FPGA programmed ({len(data)} bytes)")
        else:
            print("⚠️  CDONE is low — check connections")

    def _flash_spi_xfer(self, data, nbits=8):
        """Bit-bang SPI for Flash access.

        Flash uses SWAPPED SI/SO relative to FPGA config:
        - Flash SI ← Pi drives CFG_SO (BCM 13) → FPGA pin 67 → Flash SI
        - Flash SO → Pi reads CFG_SI (BCM 6) ← FPGA pin 68 ← Flash SO
        """
        result = 0
        for i in range(nbits - 1, -1, -1):
            GPIO.output(CFG_SO, (data >> i) & 1)   # MOSI to flash = CFG_SO
            time.sleep(0.00001)
            if GPIO.input(CFG_SI):                  # MISO from flash = CFG_SI
                result |= (1 << i)
            time.sleep(0.00001)
            GPIO.output(CFG_SCK, GPIO.HIGH)
            time.sleep(0.00001)
            GPIO.output(CFG_SCK, GPIO.LOW)
            time.sleep(0.00001)
        return result

    def _flash_mode_enter(self):
        """Swap SI/SO directions for flash access."""
        GPIO.setup(CFG_SI, GPIO.IN)         # Flash SO → Pi reads
        GPIO.setup(CFG_SO, GPIO.OUT)        # Flash SI ← Pi drives

    def _flash_mode_leave(self):
        """Restore SI/SO directions for FPGA config."""
        GPIO.setup(CFG_SI, GPIO.OUT)        # FPGA SI ← Pi drives
        GPIO.setup(CFG_SO, GPIO.IN)         # FPGA SO → Pi reads

    def flash_power_up(self):
        """Wake Flash from deep power-down (0xAB command)."""
        self._flash_mode_enter()
        self.spi_begin()
        self._flash_spi_xfer(0xAB)
        self.spi_end()
        time.sleep(0.01)

    def flash_read_id(self):
        """Read SPI Flash manufacturer/device ID."""
        self.flash_power_up()
        self.spi_begin()
        self._flash_spi_xfer(0x9F)
        mfg = self._flash_spi_xfer(0)
        dev = self._flash_spi_xfer(0)
        cap = self._flash_spi_xfer(0)
        self.spi_end()
        self._flash_mode_leave()
        size_mb = 1 << cap
        print(f"Flash: MFG=0x{mfg:02X} DEV=0x{dev:02X} SIZE={size_mb} MB")

    def flash_erase(self):
        """Bulk erase SPI Flash (0xC7) with fixed wait."""
        self.flash_power_up()
        self.spi_begin()
        self._flash_spi_xfer(0x06)  # Write enable
        self.spi_end()
        self.spi_begin()
        self._flash_spi_xfer(0xC7)  # Bulk erase
        self.spi_end()
        self._flash_mode_leave()
        print("Erasing Flash...")
        time.sleep(15)
        print("Erase complete")

    def flash_program(self, bitfile):
        """Program bitstream to QSPI Flash."""
        with open(bitfile, 'rb') as f:
            data = f.read()

        self._flash_mode_enter()
        self.flash_erase()
        self._flash_mode_enter()  # flash_erase calls _flash_mode_leave

        total = len(data)
        for addr in range(0, total, 256):
            chunk = data[addr:addr + 256]
            self.spi_begin()
            self._flash_spi_xfer(0x06)  # Write enable
            self.spi_end()
            self.spi_begin()
            self._flash_spi_xfer(0x02)  # Page program
            self._flash_spi_xfer((addr >> 16) & 0xFF)
            self._flash_spi_xfer((addr >> 8) & 0xFF)
            self._flash_spi_xfer(addr & 0xFF)
            for b in chunk:
                self._flash_spi_xfer(b)
            self.spi_end()
            if addr % 8192 == 0:
                print(f"  {addr} / {total} bytes...")
            time.sleep(0.01)

        self._flash_mode_leave()
        print(f"Flash programmed ({len(data)} bytes)")
        # Release pins so FPGA can boot from flash
        self.cleanup()
        time.sleep(0.5)
        # Check CDONE
        GPIO.setmode(GPIO.BCM)
        GPIO.setup(CFG_DONE, GPIO.IN)
        done = GPIO.input(CFG_DONE)
        GPIO.cleanup()
        if done:
            print("\u2705 FPGA booted from Flash")
        else:
            print("\u26a0\ufe0f  CDONE is low — try 'icezprog.py --reset'")

    def boot_flash(self):
        """Reset FPGA to boot from Flash.
        
        Releases all CFG pins to Hi-Z so the FPGA can drive the SPI bus
        as a master and load its configuration from the flash chip.
        """
        print("Booting from Flash...")
        # Ensure SS_B is HIGH before release (SPI Master mode)
        GPIO.setup(CFG_SS, GPIO.OUT, initial=GPIO.HIGH)
        GPIO.output(CFG_RST, GPIO.LOW)
        time.sleep(0.01)
        GPIO.output(CFG_RST, GPIO.HIGH)
        time.sleep(0.001)
        # Release ALL pins to Hi-Z — FPGA takes over SPI bus
        self.cleanup()
        time.sleep(0.5)
        # Quick check: re-init just to read CDONE
        GPIO.setmode(GPIO.BCM)
        GPIO.setup(CFG_DONE, GPIO.IN)
        done = GPIO.input(CFG_DONE)
        GPIO.cleanup()
        if done:
            print("✅ FPGA booted from Flash")
        else:
            print("⚠️  CDONE is low")


if __name__ == '__main__':
    prog = IceZeroProg()
    try:
        if len(sys.argv) < 2:
            print("Usage: icezprog.py <design.bin>        — program SRAM")
            print("       icezprog.py <design.bin> flash  — program Flash")
            print("       icezprog.py --reset            — boot from Flash")
            print("       icezprog.py --id               — read Flash ID")
            sys.exit(1)

        if sys.argv[1] == '--reset':
            prog.boot_flash()
        elif sys.argv[1] == '--id':
            prog.flash_read_id()
        elif len(sys.argv) > 2 and sys.argv[2] == 'flash':
            prog.flash_program(sys.argv[1])
        else:
            prog.prog_sram(sys.argv[1])
    finally:
        prog.cleanup()

File: examples/blinky/icezprog.c
/*
 *  icezprog — FPGA Programming Tool for ICEZero (TE0876-03, REV02)
 *
 *  Adapted from cliffordwolf/icotools examples/icezero/icezprog.c
 *  Original by Kevin M. Hubbard (blackmesalabs/ice_zero_prog)
 *
 *  GPIO pin mapping for TE0876-03 REV02 (from Trenz pinout.xlsx):
 *    CFG_DONE  = GPIO5  (P11-29)
 *    CFG_SI    = GPIO6  (P11-31)
 *    CFG_SS    = GPIO12 (P11-32)
 *    CFG_SO    = GPIO13 (P11-33)
 *    CFG_SCK   = GPIO16 (P11-36)
 *    CFG_RST   = GPIO26 (P11-37)
 *
 *  Build:  gcc -o icezprog icezprog.c -llgpio
 *  Usage:  ./icezprog design.bin    — program FPGA SRAM
 *          ./icezprog .             — program FPGA Flash (first sector)
 *          ./icezprog ..            — reset & boot from Flash
 *
 *  Requires lgpio (sudo apt install liblgpio-dev).
 *  Auto-detects gpiochip4 (Pi5) or gpiochip0 (Pi4 and earlier).
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <sys/stat.h>
#include <lgpio.h>

// ── ICEZero REV02 Pin Definitions (BCM GPIO numbers) ──
#define CFG_SS   12  // P11-32, GPIO12
#define CFG_SCK  16  // P11-36, GPIO16
#define CFG_SI    6  // P11-31, GPIO6
#define CFG_SO   13  // P11-33, GPIO13
#define CFG_RST  26  // P11-37, GPIO26
#define CFG_DONE  5  // P11-29, GPIO5

// ── lgpio handle (set in main) ──────────────────────
static int gh = -1;

// ── GPIO helpers ────────────────────────────────────
static inline void gpio_write(int pin, int val) { lgGpioWrite(gh, pin, val); }
static inline int  gpio_read(int pin)            { return lgGpioRead(gh, pin); }

// ── Flash mode: swap SI/SO directions ───────────────
// FPGA config: CFG_SI(6)=OUT, CFG_SO(13)=IN
// Flash access: CFG_SO(13)=OUT, CFG_SI(6)=IN
static void flash_mode_enter(void) {
    lgGpioFree(gh, CFG_SI);
    lgGpioFree(gh, CFG_SO);
    lgGpioClaimInput(gh, 0, CFG_SI);    // Flash SO → Pi reads
    lgGpioClaimOutput(gh, 0, CFG_SO, 0); // Flash SI ← Pi drives
}
static void flash_mode_leave(void) {
    lgGpioFree(gh, CFG_SI);
    lgGpioFree(gh, CFG_SO);
    lgGpioClaimOutput(gh, 0, CFG_SI, 0); // FPGA SI ← Pi drives
    lgGpioClaimInput(gh, 0, CFG_SO);     // FPGA SO → Pi reads
}

static uint32_t spi_xfer(uint32_t data, int nbits) {
    uint32_t rdata = 0;
    for (int i = nbits-1; i >= 0; i--) {
        gpio_write(CFG_SI, (data >> i) & 1);
        gpio_write(CFG_SCK, 1);
        if (gpio_read(CFG_SO)) rdata |= (1 << i);
        gpio_write(CFG_SCK, 0);
    }
    return rdata;
}

// ── FPGA Reset ──────────────────────────────────────
// Toggles CRESET_B. SS_B state at CRESET release determines config mode:
//   SS_B LOW  → SPI slave (SRAM programming)
//   SS_B HIGH → SPI master boot from flash
static void fpga_reset(void) {
    gpio_write(CFG_RST, 0);
    usleep(2000);
    gpio_write(CFG_RST, 1);
    usleep(2000);   // tRP: wait for FPGA internal init (1.2 ms min)
}

// ── Program SRAM ────────────────────────────────────
static void prog_sram(FILE *f) {
    printf("Programming FPGA SRAM...\n");

    // ICE40 SPI slave config: SS_B must be LOW before releasing CRESET_B
    // so the FPGA enters SPI slave config mode instead of booting from flash.
    gpio_write(CFG_SS, 0);
    fpga_reset();
    // CDONE is LOW here — FPGA is waiting for bitstream data (expected)

    // 8 dummy clocks with SS asserted
    for (int i = 0; i < 8; i++) {
        gpio_write(CFG_SCK, 0);
        gpio_write(CFG_SCK, 1);
    }

    // Send bitstream
    int byte_cnt = 0;
    for (;;) {
        int byte = getc(f);
        if (byte == EOF) break;
        for (int i = 7; i >= 0; i--) {
            gpio_write(CFG_SI, (byte >> i) & 1);
            gpio_write(CFG_SCK, 0);
            gpio_write(CFG_SCK, 1);
        }
        byte_cnt++;
    }

    // 49 dummy clocks — CDONE should go HIGH during these if config succeeded
    for (int i = 0; i < 49; i++) {
        gpio_write(CFG_SCK, 0);
        gpio_write(CFG_SCK, 1);
    }

    gpio_write(CFG_SS, 1);
    usleep(2000);
    if (gpio_read(CFG_DONE))
        printf("✅ FPGA programmed successfully (%d bytes)\n", byte_cnt);
    else
        fprintf(stderr, "⚠️  CDONE is low — programming may have failed\n");
}

// ── Flash bit-bang (SI/SO SWAPPED vs FPGA config!) ──
// For flash access: Pi → CFG_SO(13) → Flash SI, Pi ← CFG_SI(6) ← Flash SO
static uint32_t flash_spi_xfer(uint32_t data, int nbits) {
    uint32_t rdata = 0;
    for (int i = nbits-1; i >= 0; i--) {
        gpio_write(CFG_SO, (data >> i) & 1);  // MOSI to flash = CFG_SO
        gpio_write(CFG_SCK, 1);
        if (gpio_read(CFG_SI)) rdata |= (1 << i);  // MISO from flash = CFG_SI
        gpio_write(CFG_SCK, 0);
    }
    return rdata;
}

static void flash_spi_begin(void) { gpio_write(CFG_SS, 0); }
static void flash_spi_end(void)   { gpio_write(CFG_SS, 1); }
static void flash_power_up(void) {
    flash_spi_begin();
    flash_spi_xfer(0xAB, 8);  // release from deep power-down
    flash_spi_end();
    usleep(100);
}

static void flash_read_id(void) {
    flash_spi_begin();
    flash_spi_xfer(0x9F, 8);
    uint8_t mfg = flash_spi_xfer(0, 8);
    uint8_t mem = flash_spi_xfer(0, 8);
    uint8_t cap = flash_spi_xfer(0, 8);
    flash_spi_end();
    printf("Flash ID: %02X %02X %02X\n", mfg, mem, cap);
}

static void flash_write_enable(void) {
    flash_spi_begin();
    flash_spi_xfer(0x06, 8);
    flash_spi_end();
}

// Read status register (0x05); bit0 = WIP (write/erase in progress).
static uint8_t flash_read_status(void) {
    flash_spi_begin();
    flash_spi_xfer(0x05, 8);
    uint8_t s = flash_spi_xfer(0, 8);
    flash_spi_end();
    return s;
}

// Block until the flash finishes the current write/erase (WIP clears).
static void flash_wait_wip(void) {
    while (flash_read_status() & 0x01)
        usleep(1000);
}

// Whole-chip erase (0xC7). Required before programming a full bitstream:
// the previous per-sector erase only cleared the first 4 KB, so the rest of
// a ~135 KB bitstream was written into non-erased flash and corrupted.
static void flash_chip_erase(void) {
    flash_write_enable();
    flash_spi_begin();
    flash_spi_xfer(0xC7, 8);
    flash_spi_end();
    flash_wait_wip();   // chip erase takes several seconds
}

static void flash_erase_sector(uint32_t addr) {
    flash_write_enable();
    flash_spi_begin();
    flash_spi_xfer(0x20, 8);
    flash_spi_xfer(addr >> 16, 8);
    flash_spi_xfer(addr >> 8, 8);
    flash_spi_xfer(addr, 8);
    flash_spi_end();
    usleep(100000);  // 100ms for sector erase
}

static void flash_program_page(uint32_t addr, uint8_t *data, int len) {
    flash_write_enable();
    flash_spi_begin();
    flash_spi_xfer(0x02, 8);
    flash_spi_xfer(addr >> 16, 8);
    flash_spi_xfer(addr >> 8, 8);
    flash_spi_xfer(addr, 8);
    for (int i = 0; i < len; i++)
        flash_spi_xfer(data[i], 8);
    flash_spi_end();
    usleep(10000);  // 10ms for page program
}

// ── Program Flash ───────────────────────────────────
// CRESET_B is held LOW by main() before this is called,
// so the FPGA is in reset and the SPI flash bus is accessible.
static void prog_flash(FILE *f) {
    flash_mode_enter();
    flash_power_up();
    flash_read_id();

    printf("Erasing Flash (whole chip)...\n");
    fflush(stdout);
    flash_chip_erase();

    printf("Programming Flash...\n");
    uint8_t buf[256];
    int total = 0;
    for (int page = 0;; page++) {
        size_t n = fread(buf, 1, sizeof(buf), f);
        if (n == 0) break;
        flash_program_page(page * 256, buf, n);
        flash_wait_wip();
        total += n;
        if (total % 8192 == 0) { printf("  %d bytes...\n", total); fflush(stdout); }
    }
    printf("✅ Flash programmed (%d bytes)\n", total);
    flash_mode_leave();
}

// ── Release all GPIOs to Hi-Z (like Python GPIO.cleanup) ──
// For each pin: claim as INPUT with no pull, then free.
// This matches what RPi.GPIO.cleanup() does via lgpio.
static void release_all_pins(void) {
    int pins[] = {CFG_SS, CFG_SCK, CFG_SI, CFG_SO, CFG_RST, CFG_DONE};
    for (int i = 0; i < 6; i++) {
        lgGpioClaimInput(gh, 0, pins[i]);  // input, no pull (SET_PULL_NONE=0)
        lgGpioFree(gh, pins[i]);
    }
    lgGpiochipClose(gh);
    gh = -1;
}

// ── Boot from Flash ─────────────────────────────────
// Release all pins so FPGA can drive SPI bus as master.
static void boot_from_flash(void) {
    printf("Booting FPGA from Flash...\n");
    fflush(stdout);
    // Ensure SS_B HIGH at CRESET release for SPI Master mode
    gpio_write(CFG_SS, 1);
    gpio_write(CFG_RST, 0);
    usleep(10000);
    gpio_write(CFG_RST, 1);
    usleep(2000);  // 2ms for FPGA to sample mode pins
    // Release ALL pins (like Python GPIO.cleanup())
    release_all_pins();
    usleep(500000);
    // Quick check: re-open to read CDONE
    struct stat st;
    int chip = (stat("/dev/gpiochip4", &st) == 0) ? 4 : 0;
    int h = lgGpiochipOpen(chip);
    if (h >= 0) {
        lgGpioClaimInput(h, 0, CFG_DONE);
        if (lgGpioRead(h, CFG_DONE))
            printf("✅ FPGA booted from Flash\n");
        else
            fprintf(stderr, "⚠️  CDONE is low\n");
        lgGpiochipClose(h);
    }
}

// ── Main ────────────────────────────────────────────
int main(int argc, char **argv) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <bitstream.bin>   — program FPGA SRAM\n", argv[0]);
        fprintf(stderr, "       %s .                  — erase Flash & program\n", argv[0]);
        fprintf(stderr, "       %s ..                 — reset & boot from Flash\n", argv[0]);
        return 1;
    }

    // Auto-detect gpiochip: Pi5 uses chip 4, Pi4 and earlier use chip 0
    struct stat st;
    int chip = (stat("/dev/gpiochip4", &st) == 0) ? 4 : 0;
    gh = lgGpiochipOpen(chip);
    if (gh < 0) {
        fprintf(stderr, "Error: cannot open /dev/gpiochip%d: %s\n", chip, lguErrorText(gh));
        return 1;
    }

    lgGpioClaimOutput(gh, 0, CFG_SS,  1);   // SS idle HIGH
    lgGpioClaimOutput(gh, 0, CFG_SCK, 0);
    lgGpioClaimOutput(gh, 0, CFG_SI,  0);
    lgGpioClaimInput (gh, 0, CFG_SO);
    lgGpioClaimOutput(gh, 0, CFG_RST, 0);   // hold FPGA in reset
    lgGpioClaimInput (gh, 0, CFG_DONE);

    if (strcmp(argv[1], "..") == 0) {
        boot_from_flash();
    } else if (strcmp(argv[1], ".") == 0) {
        prog_flash(stdin);
        boot_from_flash();
        return 0;
    } else if (strcmp(argv[1], "..") == 0) {
        // Boot only: set SS high, pulse reset, release pins
        gpio_write(CFG_SS, 1);
        gpio_write(CFG_RST, 0);
        usleep(10000);
        gpio_write(CFG_RST, 1);
        usleep(2000);
        release_all_pins();
        usleep(500000);
        // Check CDONE
        chip = (stat("/dev/gpiochip4", &st) == 0) ? 4 : 0;
        int h = lgGpiochipOpen(chip);
        if (h >= 0) {
            lgGpioClaimInput(h, 0, CFG_DONE);
            printf(lgGpioRead(h, CFG_DONE) ? "✅ FPGA booted from Flash\n" : "⚠️  CDONE is low\n");
            lgGpiochipClose(h);
        }
        return 0;
    } else {
        FILE *f = fopen(argv[1], "rb");
        if (!f) { perror(argv[1]); lgGpiochipClose(gh); return 1; }
        prog_sram(f);
        fclose(f);
    }

    lgGpiochipClose(gh);
    return 0;
}
